### PR TITLE
enhance: Off removes all listeners for EventEmitter

### PR DESCRIPTION
**Pull request by mattheworiordan**
See https://github.com/ably/docs/issues/82

@paddybyers @SimonWoolf please review

### CODE REVIEW

**Comment by mattheworiordan** [comment]
cc @tcard FYI in regards to the discussion at https://github.com/ably/ably-ios/pull/179/files#r51544655

File: common/lib/util/eventemitter.js [hunk 1]
```diff
@@ -18,6 +18,31 @@ var EventEmitter = (function() {
...
+		for (targetListenersIndex = 0; targetListenersIndex < targetListeners.length; targetListenersIndex++) {
+			listeners = targetListeners[targetListenersIndex];
+			if (Utils.isArray(listeners)) {
+				while ((idx = Utils.arrIndexOf(listeners, listener)) !== -1) {
+					listeners.splice(idx, 1);
+				}
+			} else if (Utils.isObject(listeners)) {
+				/* events */
+				for (eventName in listeners) {
+					if (listeners.hasOwnProperty(eventName) && Utils.isArray(listeners[eventName])) {
```
Comment (paddybyers) [inline common/lib/util/eventemitter.js]: This does not delete `listeners[eventName]` if it is empty after the remove operation.

**Comment by paddybyers** [comment]
LGTM apart from the comment above.

Also, someone needs to take these tests: https://github.com/ably/realtime/blob/master/test/util/eventemitter.js and migrate them.

File: common/lib/util/eventemitter.js (hunk 1 above)
Comment (mattheworiordan) [inline common/lib/util/eventemitter.js]: Does it need to be deleted?

Comment (paddybyers) [inline common/lib/util/eventemitter.js]: Yes of course. If an app for example continuously uses new randomly generated event names and then discards the listeners, the size of the `events` object will grow indefinitely. It's easy to re-add.

Comment (mattheworiordan) [inline common/lib/util/eventemitter.js]: > Yes of course. If an app for example continuously uses new randomly generated event names and then discards the listeners, the size of the events object will grow indefinitely. It's easy to re-add.

Ah, good point, I will make that change.

**Comment by mattheworiordan** [comment]
@paddybyers fixed with https://github.com/ably/ably-js/commit/4f50db9ae189f075b87c07e8e48fbec2f3ff122c

**Comment by paddybyers** [comment]
> fixed with 4f50db9

LGTM

**Pull request merged**